src/v/cluster/partition_balancer_planner.cc

```diff
@@ -682,26 +699,30 @@ auto partition_balancer_planner::request_context::do_with_partition(
     return visitor(part);
 }
 
-void partition_balancer_planner::request_context::for_each_partition(
+ss::future<> partition_balancer_planner::request_context::for_each_partition(
   ss::noncopyable_function<ss::stop_iteration(partition&)> visitor) {
-    for (const auto& t : _parent._state.topics().topics_map()) {
-        for (const auto& a : t.second.get_assignments()) {
-            auto ntp = model::ntp(t.first.ns, t.first.tp, a.id);
+    const auto& topics = _parent._state.topics();
+    for (auto it = topics.topics_iterator_begin();
+         it != topics.topics_iterator_end();
+         ++it) {
+        for (const auto& a : it->second.get_assignments()) {
```

Comment:
I think topic table revision check on operator dereference is insufficient here, because it will happen only once at the beginning of the loop, and we need to check after each yield (due to `create_partition_cmd`).